### PR TITLE
Enable raytracer smoother for SymmetricCylinder - work in progress

### PR DESCRIPTION
**Pull request by Lathreas**
This contains a draft to make raytraced surfaces smooth near the boundaries using an arbitrary smooth falloff function. Currently the corners are not yet implemented, to be done shortly. Code cleanup will also be done.

### CODE REVIEW

File: engine/geometry/IExtendedRaytraceableSurface.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		/// </summary>
+		public float RayLength { get; set; }
+
+		/// <summary>
+		/// 	This value is <c>0.0f</c> when the ray intersects the surface. When the ray intersection is outside of
+		/// 	the surface, this value indicates the distance to the closest boundary.
+		/// </summary>
+		public float DistanceToSurface { get; set; }
+
+		public RayExtendedSurfaceIntersection(float rayLength, float distanceToSurface)
```
Comment (AdamNorberg) [inline engine/geometry/IExtendedRaytraceableSurface.cs:35]: This constructor doesn't pull its weight - I recommend removing it. C# allows struct literal initialization for structs with only public properties/fields, and then it's immediately obvious at the call site that nothing subtle is going on (because it uses brace initialization instead of invoking a constructor with parentheses).

(Note that this is different from C++, where - courtesy of initializer_list constructors - there is absolutely no guarantee that brace initialization is any less confusing than standard constructor initialization...)

File: engine/geometry/IExtendedRaytraceableSurface.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace FreedomOfFormFoundation.AnatomyEngine.Geometry
+{
+	/// <summary>
+	/// 	An interface defining the ability to raytrace a surface analytically.
+	/// </summary>
+	public struct RayExtendedSurfaceIntersection
+	{
+		/// <summary>
+		/// 	The distance from the start point of the ray to the intersection point.
```
Comment (AdamNorberg) [inline engine/geometry/IExtendedRaytraceableSurface.cs:25]: What value does this hold if the ray does not intersect the surface?

File: engine/geometry/IExtendedRaytraceableSurface.cs [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+namespace FreedomOfFormFoundation.AnatomyEngine.Geometry
+{
+	/// <summary>
+	/// 	An interface defining the ability to raytrace a surface analytically.
+	/// </summary>
+	public struct RayExtendedSurfaceIntersection
```
Comment (AdamNorberg) [inline engine/geometry/IExtendedRaytraceableSurface.cs:22]: It seems odd for this "Intersection" struct not to actually declare where (if anywhere) the intersection occurs, like RaySurfaceIntersection does.  I would kind of expect a RayExtendedSurfaceIntersection to have a RaySurfaceIntersection field and its extra field(s), and add properties to "call through" to the public properties of the embedded RaySurfaceIntersection, although this would complicate construction (so my comment below about just using brace initialization would be wrong).

File: engine/geometry/IRaytraceableSurface.cs [hunk 4]
```diff
@@ -16,6 +16,34 @@
...
+		/// 	The U coordinate on the surface at the point of intersection.
+		/// </summary>
+		public float U { get; set; }
+
+		/// <summary>
+		/// 	The V coordinate on the surface at the point of intersection.
+		/// </summary>
+		public float V { get; set; }
+
+		public RaySurfaceIntersection(float rayLength, float u, float v)
```
Comment (AdamNorberg) [inline engine/geometry/IRaytraceableSurface.cs:39]: Again, this is a good candidate for removal, brace initialization will do.

File: engine/geometry/IRaytraceableSurface.cs [hunk 5]
```diff
@@ -16,6 +16,34 @@
 
 namespace FreedomOfFormFoundation.AnatomyEngine.Geometry
 {
+	/// <summary>
+	/// 	The intersection point on a surface after a raycast.
+	/// </summary>
+	public struct RaySurfaceIntersection
```
Comment (AdamNorberg) [inline engine/geometry/IRaytraceableSurface.cs:22]: For all fields: What values does this hold if it represents a non-intersection?

File: engine/geometry/Plane.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+using System;
+using FreedomOfFormFoundation.AnatomyEngine.Calculus;
+
+namespace FreedomOfFormFoundation.AnatomyEngine.Geometry
+{
+	/// <summary>
+	/// 	A <c>Plane</c> represents a planar surface that can be deformed by a heightmap. Due to its simplicity, it is
+	/// 	raytraceable.
+	/// </summary>
+	public class Plane : IRaytraceableSurface
```
Comment (AdamNorberg) [inline engine/geometry/Plane.cs:28]: I don't like the name Plane for this, since it is not a geometric plane due to its use of heightmap deformation. Consider HeightmappedPlane, which seems more obvious to me.

File: engine/geometry/Plane.cs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		/// 	One of the two vectors defining the span and size of the plane. The v-coordinate follows this vector.
+		/// </param>
+		/// <param name="origin">
+		/// 	The starting point in 3D space of the plane. This is where the parametric coordinates (u,v) are (0,0).
+		/// </param>
+		/// <param name="heightmap">
+		/// 	The heightmap that deforms the plane in the normal direction. The normal is defined as the normalized
+		/// 	cross product between <c>VectorU</c> and <c>VectorV</c>.
+		/// </param>
+		public Plane(Vector3 origin, Vector3 vectorU, Vector3 vectorV, RaytraceableFunction1D heightmap, float heightFactor = 1.0f)
```
Comment (AdamNorberg) [inline engine/geometry/Plane.cs:52]: I feel like this is a spot where explicit objects for "vector spaces" would reduce the opportunity for bugs, fwiw. (going back to that CL where I drafted the idea out, which is still just kinda sitting.)

More immediately practical: what happens if a Plane is not uniquely defined by the provided arguments? If vectorU is some multiple of vectorV, or either is zero (which, I suppose, is just the zero multiple...), no plane is defined.

File: engine/geometry/Plane.cs [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+			heightmap1D = heightmap;
+			VectorU = vectorU;
+			VectorV = vectorV;
+			Origin = origin;
+		}
+		#endregion
+
+		#region Properties
+		// TODO: change access modifier
+		protected RaytraceableFunction1D heightmap1D;
```
Comment (AdamNorberg) [inline engine/geometry/Plane.cs:64]: That the plane is only deformable from a 1d argument rather than 2d surprises me. It's exactly what the constructor documentation says and what types are involved, so I don't think there's anything to immediately change, I'm just surprised and I'm interested in why you're using a 1d map here.

File: engine/geometry/Plane.cs [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+		private Vector3 _vectorU;
+		public Vector3 VectorU {
+			get { return _vectorU; }
+			set
+			{
+				_vectorU = value;
+				Matrix4x4 planeMatrix = new Matrix4x4(VectorU.X, VectorU.Y, VectorU.Z, 0.0f,
+													  VectorV.X, VectorV.Y, VectorV.Z, 0.0f,
+													  Normal.X,  Normal.Y,  Normal.Z,  0.0f,
+													  0.0f,      0.0f,      0.0f,      1.0f);
```
Comment (AdamNorberg) [inline engine/geometry/Plane.cs:76]: Here and everywhere you're constructing these planeMatrix objects, something has gone wrong with indentation. Are you indenting with tabs, using a tab stop size different from Github's expectation?

Either indent with spaces or indent _only_ with tabs, including using them as column breaks, and break after the opening parenthesis to avoid needing to try to align tabbed continuation rows with the opening parenthesis (which is in a character-fixed position).

FWIW, I prefer tab indentation _when it is consistently and correctly used_, which means strictly forbidding the "first line of arguments inline with opening paren, subsequent lines indented straight under it" pattern because it's impossible to do with varied tab stop sizes. My reason for preferring it is because it's useful for editors that use proportional fonts and use font size as part of syntax highlighting - I think Source Insight is the only popular one, admittedly, but I wish it was more common.

tl;dr: indentation broken, plz fix.

File: engine/geometry/Plane.cs [hunk 10]
```diff
@@ -0,0 +1,157 @@
...
+			{
+				// First calculate the parameters u and v on the surface of the Plane at which the ray intersects.
+				float u = r0.X + i * r1.X;
+				float v = r0.Y + i * r1.Y;
+
+				// t must be within the bounds of the Plane. t ranges from 0.0 to 1.0, so if it is outside
+				// of this region, the ray does not intersect the cylinder.
+				if (!(u >= 0.0f && u <= 1.0f) || !(v >= 0.0f && v <= 1.0f))
+				{
+					break;
```
Comment (AdamNorberg) [inline engine/geometry/Plane.cs:137]: Is the output of heightmap1D.SolveRaytracePlanar sorted? If so, why continue iteration after finding anything that works, since anything later in the list will be further away? If not, why is this `break` rather than `continue`?

File: engine/geometry/SymmetricCylinder.cs [hunk 11]
```diff
@@ -168,24 +168,25 @@ public float RayIntersect(Ray ray)
...
-				// Now we have the parameter t on the surface of the SymmetricCylinder at which the ray intersects.
-				
+				// First calculate the parameter t on the surface of the SymmetricCylinder at which the ray intersects.
+				float t = rescaledRay.Z + i*newDirection.Z;
+
+				// t must be within the bounds of the SymmetricCylinder. t ranges from 0.0 to 1.0, so if it is outside
+				// of this region, the ray does not intersect the cylinder.
+				if (!(t >= 0.0f && t <= 1.0f))
+				{
+					break;
```
Comment (AdamNorberg) [inline engine/geometry/SymmetricCylinder.cs:182]: As above: why not `continue`?

File: engine/geometry/SymmetricCylinder.cs [hunk 12]
```diff
@@ -196,12 +197,89 @@ public float RayIntersect(Ray ray)
 				// otherwise discard:
 				if ( phi > StartAngle.GetValueAt(t) && phi < EndAngle.GetValueAt(t) && i >= 0.0f)
 				{
-					minimum = Math.Sign(i)*(float)Math.Min(Math.Abs(minimum), Math.Abs(i));
+					if (Math.Abs(i) < Math.Abs(minimum.RayLength))
```
Comment (AdamNorberg) [inline engine/geometry/SymmetricCylinder.cs:200]: As above, this check should be reversed and first.

File: engine/geometry/Plane.cs [hunk 13]
```diff
@@ -0,0 +1,157 @@
...
+				}
+
+				// Determine if the ray is inside the pie-slice of the cylinder that is being displayed,
+				// otherwise discard:
+				if (i >= 0.0f)
+				{
+					if (Math.Abs(i) < Math.Abs(minimum.RayLength))
```
Comment (AdamNorberg) [inline engine/geometry/Plane.cs:144]: If the output isn't sorted, I feel like the inversed version of this check should be first and call `continue` if it's not met, since it's a cheap float vs. float comparison that filters whether this iteration can possibly do anything (we won't have a new minimum if this intersection is further away, after all).

File: engine/geometry/Plane.cs [hunk 14]
```diff
@@ -0,0 +1,157 @@
...
+				float v = r0.Y + i * r1.Y;
+
+				// t must be within the bounds of the Plane. t ranges from 0.0 to 1.0, so if it is outside
+				// of this region, the ray does not intersect the cylinder.
+				if (!(u >= 0.0f && u <= 1.0f) || !(v >= 0.0f && v <= 1.0f))
+				{
+					break;
+				}
+
+				// Determine if the ray is inside the pie-slice of the cylinder that is being displayed,
```
Comment (AdamNorberg) [inline engine/geometry/Plane.cs:140]: What do cylinders have to do with this? (I infer this is a bad copy-paste from the Cylinder versions of this logic)

File: engine/geometry/Plane.cs (hunk 13 above)
Comment (AdamNorberg) [reply engine/geometry/Plane.cs:144]: Additionally, the Math.Abs(i) call is not required since we already know that i is mostly nonnegative. (It may be -0.0 but nothing bad happens if it's negative zero so the existing check is "nonnegative enough".)

File: engine/geometry/SymmetricCylinder.cs [hunk 15]
```diff
@@ -196,12 +197,89 @@ public float RayIntersect(Ray ray)
...
+		/// <inheritdoc />
+		public RayExtendedSurfaceIntersection ExtendedRayIntersect(Ray ray)
+		{
+			RaySurfaceIntersection intersection = RayIntersect(ray);
+
+			if (Single.IsNaN(intersection.RayLength) == false && Single.IsInfinity(intersection.RayLength) == false)
+			{
+				return new RayExtendedSurfaceIntersection(intersection.RayLength, 0.0f);
+			}
+			else
```
Comment (AdamNorberg) [inline engine/geometry/SymmetricCylinder.cs:223]: This is an unpleasantly wedge-shaped function! When an if block ends in a `return` on all codepaths (as a one-line return block clearly does!), skip the `else` - the only way execution continues is implicitly `else`, so you can get rid of a lot of indentation by just removing all the `else`s subsequent to `if...return` blocks. (that is, remove the else, remove the curly brackets, outdent the code so it runs "unconditionally" - except execution of the function will already have halted if the `if` was taken.)

File: engine/geometry/Plane.cs (hunk 8 above)
Comment (Lathreas) [reply engine/geometry/Plane.cs]: It's meant to be a 1D deformation to give a corrugated effect, so that whatever shape the edge of a HingeJoint is becomes extended and smoothly falls off (if that makes sense, otherwise I can explain the method with a few pictures - probably good idea anyway for documentation purposes). Indeed, from the name it feels like it should be 2D, so that is indeed confusing. I've changed the name accordingly (see comment above).

File: engine/geometry/Plane.cs (hunk 10 above)
Comment (Lathreas) [reply engine/geometry/Plane.cs]: Oh, good catch! Should be `continue` indeed. Fixed.
